### PR TITLE
[ZEPPELIN-5644] Reduce redundancy during the interpreter execution process

### PR DESCRIPTION
**Pull request by saLeox**
### What is this PR for?
Optimize the interpreter execution process from the following aspects:
1. Close the interpreter executor thread when restart interpreter;
2. Only close opened interpreter when restart interpreter and create non-existent interpreter when create Interpreter;
3. For spark interpreter, reuse the static innerScalaInterpreter class, in the context of multi sub threads, which is only loaded by the first interpreter;
4. When close spark interpreter, set the inner objects (SparkZeppelinContext and sparkILoop) to empty.

### What type of PR is it?
[Improvement]

### Todos
* [ ] - Task

### What is the Jira issue?
* Open an issue on Jira https://issues.apache.org/jira/browse/ZEPPELIN-5644

### How should this be tested?
* Covered by existing unit test

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? No

### CODE REVIEW

**Comment by jongyoul** [comment]
@saLeox Thank you for your contribution. BTW, could you please add some tests for this feature? I think your implementation looks reasonable but it would be better we have tests.

**Comment by saLeox** [comment]
> @saLeox Thank you for your contribution. BTW, could you please add some tests for this feature? I think your implementation looks reasonable but it would be better we have tests.
Hi @jongyoul, thanks for your reply, the modification I made are covered by some existing junit method, like SparkInterpreterTest.testSparkInterpreter() or RemoteInterpreterServerTest.testInterpreter(), should I add some specified cases in those methods?

File: zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/remote/RemoteInterpreterServer.java
Comment (Reamer) [inline zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/remote/RemoteInterpreterServer.java:493]: Is this cast safe? I would prefer an Instanceof test before making this cast.

Comment (saLeox) [reply zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/remote/RemoteInterpreterServer.java:493]: Thanks for your suggestion, Reamer, I will add the condition.

**Comment by saLeox** [comment]
Hi all, I add the unit tests for the features 1 to 3, pls help review.

**Comment by saLeox** [comment]
@Reamer Hi, do you have time to have another round of review?

**Comment by zjffdu** [comment]
@saLeox Sorry for the late response, the CI is unstable recently. We are working on it https://github.com/apache/zeppelin/pull/4266
So we need to rerun the CI before merging this PR.

**Comment by zjffdu** [comment]
@saLeox CI issue is fixed, could you rebase this PR to the latest master?

**Comment by saLeox** [comment]
@zjffdu Sure, thanks!

**Comment by saLeox** [comment]
@zjffdu Hi, any further feedback or concern regarding this PR? The integrated test seems running well after the rebase.

File: spark/interpreter/src/main/java/org/apache/zeppelin/spark/SparkInterpreter.java
Comment (zjffdu) [inline spark/interpreter/src/main/java/org/apache/zeppelin/spark/SparkInterpreter.java:71]: Move it together with other static variables.

File: spark/interpreter/src/test/java/org/apache/zeppelin/spark/SparkInterpreterTest.java
Comment (zjffdu) [inline spark/interpreter/src/test/java/org/apache/zeppelin/spark/SparkInterpreterTest.java:20]: Not necessary

**Comment by zjffdu** [comment]
@saLeox Sorry for late response, just a few minor comments

**Comment by Reamer** [comment]
> @Reamer Hi, do you have time to have another round of review?

I'm afraid I can't do a code review here, as I'm not familiar with this part of the code.

**Comment by saLeox** [comment]
@zjffdu  Thanks for the comments, I will commit and rebase again accordingly.

**Review by zjffdu — APPROVED** [review]

**Comment by zjffdu** [comment]
@saLeox Could you do a rebase and trigger CI again?

**Comment by saLeox** [comment]
@zjffdu Done, and pls help approve the new CI

**Comment by zjffdu** [comment]
CI is passed, will merge if no more comment

**Pull request merged**
